File: 648-replace-words/648-replace-words.java
class Solution {
    
    TrieNode root;
    
    public String replaceWords(List<String> dictionary, String sentence) {
        root = new TrieNode();
        buildTrie(dictionary);
        String[] words = sentence.split(" ");
        
        for(int i = 0; i < words.length; i++)
            words[i] = findRoot(words[i]);
        
        return String.join(" ", words);
    }
    
    private String findRoot(String word){
        TrieNode temp = root;
        StringBuilder sb = new StringBuilder();
        
        for(int i = 0; i < word.length(); i++){
            int currCharIdx = word.charAt(i) - 'a';
            if(temp.isWord) break;
            if(temp.children[currCharIdx] != null){
                sb.append(word.charAt(i));
                temp = temp.children[currCharIdx];
            }else break;
        }
        
        return (sb.isEmpty() || !temp.isWord) ? word : sb.toString();
    }
    
    private void buildTrie(List<String> words){
        for(String word : words) insert(word);
    }
    
    private void insert(String word){
        TrieNode temp = root;
        for(int i = 0; i < word.length(); i++){
            int currCharIdx = word.charAt(i) - 'a';
            if(temp.children[currCharIdx] == null){
                temp.children[currCharIdx] = new TrieNode();
            }
            temp = temp.children[currCharIdx];
        }
        temp.isWord = true;
    }
    
}

class TrieNode{
    
    boolean isWord;
    TrieNode[] children;
    
    public TrieNode(){
        children = new TrieNode[26]; // lowercase alphabates 
    }
}

File: 648-replace-words/648-replace-words.py
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        
        root = TrieNode() 
        
        def insert(word):
            curr = root 
            for char in word:
                idx = ord(char) - ord('a')
                if curr.children[idx] is None:
                    curr.children[idx] = TrieNode() 
                curr = curr.children[idx] 
            curr.isEnd = True 
            
        def find(word):
            curr = root 
            curr_word = []
            for char in word: 
                if curr.isEnd: 
                    return "".join(curr_word)
                idx = ord(char) - ord('a') 
                if curr.children[idx] is None:
                    return word
                curr_word.append(char)
                curr = curr.children[idx] 
            return word 
        
        ans = []
        
        for word in dictionary:
            insert(word)
        
        for word in sentence.split():
            ans.append(find(word))
        
        return " ".join(ans)
                
        
        
class TrieNode: 
    def __init__(self): 
        self.children = [None] * 26 
        self.isEnd = False 